### DISCUSSION TITLE
Questions about the code level of AppDefinition

### DISCUSSION THREAD

**Discussion by wyhdil** [eclipse-theia/theia-cloud#243] [Q&A] [answered]
Hello:
  I have some questions and I wonder if you can help me solve them.

AppDefinition should be used to define the IDE image and required resources, right?

![781f2ac8829f3f3a2e968b5bcb30dd89](https://github.com/eclipsesource/theia-cloud/assets/69418800/b4cf22c5-ac23-4740-9528-2aa86b769951)
But I can't find where the assignment is made.

![caeb75f86563031d793bdb55adba0eba](https://github.com/eclipsesource/theia-cloud/assets/69418800/6c27ab98-3a62-4f1d-97d1-d1ee8a5348fb)
It looks like there are corresponding fields here.But I'm not sure if it's right.

I would like to ask where I can see an explanation of the code directory structure or a more detailed explanation of the code level.

Thanks！

**Comment by sgraband** [comment]
Hey @wyhdil,

which assignment are you talking about?
The data in the second screenshot are Kubnernetes Resources, that are being applied to the cluster. Most likely via the service.
The classes in the first screenshot are the POJOs for those resources in Java. We use those to handle the created resources with the operator. The operator watches the Cluster for those resources and transforms them to the POJO using [fabric.io's kubernetes client](https://github.com/fabric8io/kubernetes-client).

I hope this answers your question. 
Please note, that this is a free support forum and we cannot provide full tutorials/workshops in this forum. If you want more detailed coaching about Theia Cloud you could consider booking one of our [developer support packages](https://eclipsesource.com/services/developer-support/).

**Reply by wyhdil** [reply]
Hello, thank you very much for your previous answer! ! !  Now I still have some questions

I mean I don't know how the IDE's pods are created. Theoretically, this pod should be created through the deployment.yaml file, but I don't know how to create this file. 

For example, how is limitsCpu resource assigned?If I want to change the value of the limitsCpu resource and use it as a variable for user customization, what should I do?

**Reply by sgraband** [reply, marked as answer]
The "deployments" in your screenshots are not deployments, but Custom Resources (CRs). You can manually apply a CR using `kubectl apply -f <fileName>`. In a standard Theia Cloud deployment this would be done via the [service](https://github.com/eclipsesource/theia-cloud/tree/main/java/service/org.eclipse.theia.cloud.service) that offers endpoints for requesting a new session (deployment).

Once the resource is applied to the cluster (it doesn't matter if that happens manually or via the service) the operator handles that resource and creates the deployment with the information of the CR. This is done, by replacing the placeholders in [this file](https://github.com/eclipsesource/theia-cloud/blob/main/java/operator/org.eclipse.theia.cloud.operator/src/main/resources/templateDeployment.yaml) and then applying that file.

The replacements are happening [here](https://github.com/eclipsesource/theia-cloud/blob/main/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/handler/impl/DefaultDeploymentTemplateReplacements.java).

Hope this helps.

**Reply by wyhdil** [reply]
Okay, thank you very much for your patient answer！ I will take a closer look at the implementation process here.

**Reply by wyhdil** [reply]
@sgraband There is something I need to confirm, do you mean to replace [that](https://github.com/eclipsesource/theia-cloud/blob/main/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/handler/impl/DefaultDeploymentTemplateReplacements.java) file with the contents of [this](https://github.com/eclipsesource/theia-cloud/blob/main/demo/k8s/appdefinitions/theia.yaml) file?

If this is the case, what is the process of implementing this process at the code level?（The main reason is that I don’t know how [this](https://github.com/eclipsesource/theia-cloud/blob/main/demo/k8s/appdefinitions/theia.yaml) file is imported and used.）

I would be very grateful if you could answer!

**Reply by sgraband** [reply]
Yes. The operator [watches all Custom Resources](https://github.com/eclipsesource/theia-cloud/blob/main/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/SpecWatch.java) that are related to Theia Cloud on the cluster. Once he finds such a resource it is transformed into a POJO and in the replacement class it is the [AppDefinition object](https://github.com/eclipsesource/theia-cloud/blob/685e1eca2f3b78b2f4882a54f03ffd6076f4a1e9/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/handler/impl/DefaultDeploymentTemplateReplacements.java#L118).

**Reply by wyhdil** [reply]
OK, thanks!

**Reply by wyhdil** [reply]
Hello, I still have some questions to ask. @sgraband 
I sorted out the process of resources required to create an IDE pod above. 

First, I want to confirm if it is correct.
The [theia.yaml](https://github.com/eclipsesource/theia-cloud/blob/main/demo/k8s/appdefinitions/theia.yaml) will be applied to the cluster by the service and become a resource. [SpecWatch.java](https://github.com/eclipsesource/theia-cloud/blob/main/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/SpecWatch.java) will watch this resource, and then transform it into a [POJO](https://github.com/eclipsesource/theia-cloud/blob/main/java/common/org.eclipse.theia.cloud.common/src/main/java/org/eclipse/theia/cloud/common/k8s/resource/AppDefinitionSpec.java) of App Definition class. Then the operator replaces the App Definition part of this [templateDeployment.yaml](https://github.com/eclipsesource/theia-cloud/blob/main/java/operator/org.eclipse.theia.cloud.operator/src/main/resources/templateDeployment.yaml) through this [DefaultDeploymentTemplateReplacements.java](https://github.com/eclipsesource/theia-cloud/blob/main/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/handler/impl/DefaultDeploymentTemplateReplacements.java).

If this process is correct, I would like to ask, is this process already executed in the first step?
I now want to allow users to select the size of the cpu resource when logging in, and then create IDE pods based on customized CPU resource.However, judging from the [architecture diagram](https://github.com/eclipsesource/theia-cloud/blob/main/doc/docs/theia.cloud.png), the sequence of the entire process is difficult to achieve the above goals.Because the App Definition CR has been completed in the first step, and the user login process is after that.

So I'm wondering if there's a way to accomplish my goal?

**Reply by sgraband** [reply]
Your assumptions are correct. However, the replacements are only done as soon as the user requests a session and a Session CR ist created. The session CR will then point to the AppDefinition CR that you want to start a session of.
Regarding your problem you probably have two options:
1. If you only have a limited amount of options for the CPU resource (e.g. only 1vCPU or 2vCPU) you coul just create two app defintions that have the values. The user can then select which AppDefinition he would like to use.
2. If the options for the CPU resources can be anything from 0.1 to lets say 10. This would probably not be feasible. The solution would then be to extend Session CRD to have a CPU field and adjust the Operator so that it takes the value from the session and not from the appdefinition, when it replaces the values.

hope this helps.

**Reply by wyhdil** [reply]
What a coincidence, I'm trying to make session CR expand the cpu's fields. After seeing your answer, I am even more convinced that this is possible! Thank you!

**Reply by wyhdil** [reply]
@sgraband Hello,I appreciate your patience with my previous questions. However, I still have a few more queries. Would you mind assisting me with them?  I encountered some problems when extending Session CRD.

Judging from the request initiated by landing-page, the service should start working from [this page](https://github.com/eclipsesource/theia-cloud/blob/main/java/service/org.eclipse.theia.cloud.service/src/main/java/org/eclipse/theia/cloud/service/RootResource.java#L58). This seems to be related to creating a session. So I passed the cpu and mem in the request body as parameters to the launchWorkspaceSession function.
![5247540ba05da5b8409ef2330368cd95](https://github.com/eclipsesource/theia-cloud/assets/69418800/55531a68-2936-4f9f-be2e-d4b66eb390bc)

At the same time, I added cpu and mem related fields to [sessionSpec.java](https://github.com/eclipsesource/theia-cloud/blob/main/java/common/org.eclipse.theia.cloud.common/src/main/java/org/eclipse/theia/cloud/common/k8s/resource/SessionSpec.java).

![9942a8b3642dc5ff3e5ded33566f48bc](https://github.com/eclipsesource/theia-cloud/assets/69418800/b09983bf-0650-4de8-b5e3-b14284acae4e)
I want to know where is the part used to create session CR. Is it SessionSpec.java that creates session CR? If so, is the modification here enough to create session CR containing cpu and mem fields?

When expanding the fields, I also found another question, what is [this module](https://github.com/eclipsesource/theia-cloud/tree/main/java/service/org.eclipse.theia.cloud.service/src/main/java/org/eclipse/theia/cloud/service/session) used for?

![1dd53bc4bb45274175dfef8497b7bdf4](https://github.com/eclipsesource/theia-cloud/assets/69418800/20c1e0b6-1f5c-4fc8-83b7-a0dee302a3ad)
What is this part doing? I mean what is its function? Is this related to creating session CR? This seems to be receiving the initiated /service/session request, but I can't find who initiated this request.

Thank you in advance for taking the time to answer my question！

**Reply by sgraband** [reply]
Your assumptions are all correct and your on a good way i believe. Regarding the service, it has multiple endpoints which you can take a look at in this [file](https://github.com/eclipsesource/theia-cloud/blob/main/doc/docs/openapi.json).

What might be confusing is the difference between LaunchSession and StartSession.
LaunchSession is simply a shortcut for demo application, while StartSession offers more flexibility.

You need to check, what your landing page (if you are using one) is using. This way you can find out the request that is being sent so you can debug/check on what you need to change in that function to add your custom parameters.

Hope this helps!

**Reply by wyhdil** [reply]
Hello,I have encountered some difficulties again and would like to ask what's going  @sgraband 

I encountered some difficulties again. I wanted to ask about the issue. I made modifications to the landing-page, service, and operator, and then built new images. However, when I tested them and tried to start the IDE, it didn't work properly. I suspected that there might be some errors in the code I modified, so I obtained the source code for[ the tag ](https://github.com/eclipsesource/theia-cloud/tree/0.8.1.MS3) from the official website. I directly built new images for the landing-page and service from that code and applied them to the cluster. However, even with these changes, the IDE still didn't start properly and returned a 473 status code. This led me to suspect that there might be some issues with the service code itself, causing it to malfunction.
![7b4ea8cdcffd17129f8a111c9f93b49d](https://github.com/eclipsesource/theia-cloud/assets/69418800/83155b89-3f76-42c9-8dff-6b4f05238080)


To validate my assumption, I replaced the service image with theiacloud/theia-cloud-service:0.8.1.OSWeek23-v1, and this time the IDE started successfully. Therefore, I believe that only the service source code corresponding to the theiacloud/theia-cloud-service:0.8.1.OSWeek23-v1 image is correct, while the service code on the official website might not be ready yet. Alternatively, could you provide me with the source code for the theiacloud/theia-cloud-service:0.8.1.OSWeek23-v1 version? I couldn't find this specific version in the tags on the official website.

I hope you can provide an answer. Thank you!

**Reply by sgraband** [reply]
You should make sure that all the components your are using are on the same version. Otherwise there might be inconsistencies between the versions. `theiacloud/theia-cloud-service:0.8.1.OSWeek23-v1` is a version used for internal development only. Therefore, it does not have documentation. Could you just try to go to the next minor version [`0.8.1.MS2`](https://github.com/eclipsesource/theia-cloud/commit/a6b63883982e46c9236c3330b1e9f01250db9458) or the latest version `0.8.1.MS3` for all of your components? Also use this version as a baseline for your adaptations.

**Reply by wyhdil** [reply]
OK!  I will try the 0.8.1.MS2 version now!

**Reply by wyhdil** [reply]
Hello, I made modifications to the project based on the custom CPU and memory resource allocation approach [we discussed earlier](https://github.com/eclipsesource/theia-cloud/discussions/243#discussioncomment-7379301). However, I've noticed that my changes are not taking effect in the cluster. Can you please help me take a look?

![2c0d6d4758714892b5c721e745f8c44b](https://github.com/eclipsesource/theia-cloud/assets/69418800/4d810a67-a723-4c32-b50a-8611205acd01)
I extended the parameter list for the `launchWorkspaceSession` method and passed the CPU and memory values that were sent from the landing page to the service, which are referred to as `request.cpu` and `request.mem` here.


![43c1296ad20132a02c286dfbe2ca3e1f](https://github.com/eclipsesource/theia-cloud/assets/69418800/f27cfec1-d3cb-4795-bf40-06e75adb2bee)
![0cd646954b6e3535965d85d2819a0dc9](https://github.com/eclipsesource/theia-cloud/assets/69418800/7ac72f50-46ba-436e-8a53-33261664dcfe)
Then I added these fields to the `SessionSpec`. In theory, these fields should be used to define the session custom resource (CR) related attributes.
<img width="2555" alt="image" src="https://github.com/eclipsesource/theia-cloud/assets/69418800/51edbf52-c168-499e-a73e-cde81ee95794">
And within the `DefaultDeploymentTemplateReplacements`, I made changes to the template replacement method to instantiate the session CR using the newly added `cpu` and `mem` from the session.

However, I have noticed that doing this doesn't seem to take effect.
![53b8cb27f3bf246a793622f6ce112d37](https://github.com/eclipsesource/theia-cloud/assets/69418800/26b82318-07e6-488b-b150-2b1f2d3cb722)
I added a marker in the replacement process, and if these operations were executed, it should print out some information. However, upon reviewing the service logs, I didn't see any of this information printed, so I suspect that the modifications I made are not taking effect.

<img width="872" alt="image" src="https://github.com/eclipsesource/theia-cloud/assets/69418800/061e28a7-4f6e-472f-9ee5-6de5825d97f4">
I also checked the session custom resources in the cluster, and I didn't find the `cpu` and `mem` fields added there as well.

So I'm suspecting that the previous approach might be flawed. Can you provide some suggestions on how to properly extend the `cpu` and `mem` fields in the session?  
Thanks！

**Reply by sgraband** [reply]
Can you verify, that the new images are indeed being used in the pods?

**Reply by wyhdil** [reply]
Oh, right, in the previous version, the image didn't get applied correctly to the cluster, and I resolved that issue. Now, a new problem has emerged.

When I try to launch the IDE in the browser, I'm getting a 504 status code error.
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/9c16a72a-60ed-45d7-bcd9-f253ee136fc4)
So, I checked the log information for both the service and operator pods.
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/5f87cb0b-a263-455f-8680-5957fc881930)
Here is the log information for the service pod. The errors here appear to be the reason for the 504 status code.

![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/cbec0c62-d241-4b15-a9bb-cdc83992e71f)
Here is the log information for the operator pod. I suspect that the errors here are likely the key issues. Can you provide some tips for solving this issue?

**Reply by sgraband** [reply]
I am sorry, but we cannot assist you with debugging your custom code. The error message shows you directly, where there is a NullPointerExceptions so you should be able to debug the problem.

**Reply by wyhdil** [reply]
Okay, thank you very much for patiently answering all the time!
